### PR TITLE
feature:Use etcd to manager supernode ha cluster(HA first part)

### PR DESCRIPTION
**Pull request by yunfeiyanggzq**
Signed-off-by: yunfeiyangbuaa <yunfeiyang@buaa.edu.cn>

### Ⅰ. Describe what this PR did
to implement supernode ha,we use etcd to manager supernode ha cluster,every supernode should connects etcd and send suopernode information to etcd.other supernode will gets this information from etcd.supernode's on and off will be notified  to other supernode.supernode will register a supernode peer when it find a new supernode is on in this cluster 
### Ⅴ. Special notes for reviews
my pr will be submitted as follows

- implement a ha_mgr  to implement active/standby switch-----done
- add a cmd config to config ha ,use ha_mgr to change supernode status between active and standby
- add new interface to ha_mgr(active supernode send a copy of  all the dfget's request to satndby supernode to implement state synchronization )
- modify dfget's code(dfget can go on it's work after the active supernode is off)
- implement the vip

### CODE REVIEW

**Comment by Starnop** [comment]
Could you please  add a simple usage document?

File: supernode/daemon/mgr/ha/etcd_tool.go [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+//WatchActiveChange is the progress to watch the etcd,if the value of key /lock/active changes,supernode will be notified.
+func (etcd *EtcdMgr) WatchActiveChange(messageChannel chan string) {
+	var watchStartRevision int64
+	watcher := clientv3.NewWatcher(etcd.client)
+	watchChan := watcher.Watch(context.TODO(), "/lock/active", clientv3.WithRev(watchStartRevision))
+	for watchResp := range watchChan {
+		for _, event := range watchResp.Events {
+			switch event.Type {
+			case ActiveSupernodeChange:
+				messageChannel <- "string(event.Kv.Value)"
```
Comment (Starnop) [inline supernode/daemon/mgr/ha/etcd_tool.go]: why do you add a pair of quotes here?

Comment (yunfeiyangbuaa) [reply supernode/daemon/mgr/ha/etcd_tool.go]: Oh,it is my wrong.

File: supernode/daemon/mgr/ha/tool.go [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+	TryBeActive() (bool, string, error)
+
+	//ActiveResureItsStatus will keep to monitor to ensure this itself is still a active supernode now.
+	ActiveResureItsStatus()
+
+	//ActiveKillItself abandon the active status and the active supernode become standby supernode.
+	ActiveKillItself() bool
+
+	//CloseTool close the tool.
+	CloseTool() error
```
[the remark this answers is not in the document]
Comment (yunfeiyangbuaa) [reply supernode/daemon/mgr/ha/tool.go]: I am agreed with you

File: supernode/daemon/mgr/ha/etcd_tool.go [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+	keepRespChan, err := lease.KeepAlive(context.TODO(), leaseResp.ID)
+	etcd.leaseKeepAliveRsp = keepRespChan
+	if err != nil {
+		logrus.Errorf("failed to create etcd.leaseKeepAliveRsp: %v", err)
+	}
+	etcd.leaseResp = leaseResp
+	//if the lock is available,get the lock.
+	//else read the lock
+	txn := kv.Txn(context.TODO())
+	txn.If(clientv3.Compare(clientv3.CreateRevision("/lock/active"), "=", 0)).
```
[the remark this answers is not in the document]
Comment (yunfeiyangbuaa) [reply supernode/daemon/mgr/ha/etcd_tool.go]: I  will make  it a  function  parameter

File: supernode/daemon/mgr/ha/etcd_tool.go [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+	if !txnResp.Succeeded {
+		_, err = lease.Revoke(context.TODO(), leaseResp.ID)
+		return false, string(txnResp.Responses[0].GetResponseRange().Kvs[0].Value), err
+	}
+	return true, etcd.hostIP, nil
+}
+
+//ActiveKillItself cancels the renew of lease.
+func (etcd *EtcdMgr) ActiveKillItself() bool {
+	_, err := etcd.client.Revoke(context.TODO(), etcd.leaseResp.ID)
```
Comment (Starnop) [inline supernode/daemon/mgr/ha/etcd_tool.go]: I think use the following code will be better.

if _, err := etcd.client.Revoke(context.TODO(), etcd.leaseResp.ID); err != nil {
logrus.Errorf("failed to cancel a etcd lease: %v", err)
return false
}

File: supernode/daemon/mgr/ha/manager.go [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+//GiveUpActiveStatus give up its active status because of unhealthy.
+func (ha *Manager) GiveUpActiveStatus() bool {
+	return ha.tool.ActiveKillItself()
+}
+
+//StandbyToActive change the status from standby to active.
+func (ha *Manager) standbyToActive() {
+	if ha.nodeStatus == constants.SupernodeUseHaStandby {
+		ha.nodeStatus = constants.SupernodeUseHaActive
```
Comment (Starnop) [inline supernode/daemon/mgr/ha/manager.go]: return here and you will not need to add a else code block.

File: supernode/daemon/mgr/ha/manager.go [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+	} else {
+		logrus.Warnf("%s is already active,can't set it active again", ha.advertiseIP)
+	}
+}
+
+//ActiveToStandby  change the status from active to standby.
+func (ha *Manager) activeToStandby() {
+	if ha.nodeStatus == constants.SupernodeUseHaActive {
+		ha.nodeStatus = constants.SupernodeUseHaStandby
```
Comment (Starnop) [inline supernode/daemon/mgr/ha/manager.go]: So do with the above.

File: supernode/daemon/mgr/ha/manager.go [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+	}
+}
+
+//TryStandbyToActive try to change the status from standby to active.
+func (ha *Manager) tryStandbyToActive(change chan int) {
+	is, ip, err := ha.tool.TryBeActive()
+	if err != nil {
+		logrus.Errorf("failed to try to change standby status to active status")
+	}
+	if is == true {
```
Comment (Starnop) [inline supernode/daemon/mgr/ha/manager.go]:
```
if !is {
		logrus.Infof("the other supernode %s obtain the active supernode status,keep watch on it", ip)
		change <- constants.SupernodeUseHaStandby
		return
	}
	
	ha.standbyToActive()
	logrus.Infof("%s obtain the active supernode status", ha.advertiseIP)
	change <- constants.SupernodeUseHaActive
	ha.tool.ActiveResureItsStatus()
	ha.activeToStandby()
	logrus.Infof("%s finishes the active supernode status", ha.advertiseIP)
```

BTY, I don't think the variable name `is` is a good name. Maybe you can use a more meaningful instead.

**Comment by Starnop** [comment]
In addition, we should follow the following conventions when recording a log:
+ record the key info that represents who is performing the operation
+ use `success to do something` or `failed to do something` log format

File: supernode/daemon/mgr/ha/etcd_tool.go [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+
+//ObtainActiveInfo obtain the active supernode's information from etcd.
+func (etcd *EtcdMgr) ObtainActiveInfo(key string) (string, error) {
+	kv := clientv3.NewKV(etcd.client)
+	var (
+		getRes *clientv3.GetResponse
+		err    error
+	)
+	if getRes, err = kv.Get(context.TODO(), key); err != nil {
+		logrus.Errorf("failed to get the active supernode's info: %v", err)
```
[the remark this answers is not in the document]
Comment (yunfeiyangbuaa) [reply supernode/daemon/mgr/ha/etcd_tool.go]: but there are only two possibilities of the values

File: supernode/daemon/mgr/ha/etcd_tool.go [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+}
+
+//WatchActiveChange is the progress to watch the etcd,if the value of key /lock/active changes,supernode will be notified.
+func (etcd *EtcdMgr) WatchActiveChange(messageChannel chan string) {
+	var watchStartRevision int64
+	watcher := clientv3.NewWatcher(etcd.client)
+	watchChan := watcher.Watch(context.TODO(), "/lock/active", clientv3.WithRev(watchStartRevision))
+	for watchResp := range watchChan {
+		for _, event := range watchResp.Events {
+			switch event.Type {
```
[the remark this answers is not in the document]
Comment (yunfeiyangbuaa) [reply supernode/daemon/mgr/ha/etcd_tool.go]: but there are only two possibilities of the values

**Comment by yunfeiyangbuaa** [comment]
@Starnop Thanks for your review.I have modified my code with your advice

File: supernode/daemon/mgr/ha/etcd_tool.go [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+const (
+	//ActiveSupernodeOFF means there is no active supernode.
+	ActiveSupernodeOFF = ""
+	//ActiveSupernodeChange means active supernode change to standby supernode because of unhealthy.
+	ActiveSupernodeChange = 0
+	//ActiveSupernodeKeep means active supernode is health.
+	ActiveSupernodeKeep = 1
+	//etcdTimeOut is the etcd client's timeout second
+	etcdTimeOut = 10 * time.Second
+	//activeKeyPreFIx is the keyPrefix of active supernode
```
Comment (Starnop) [inline supernode/daemon/mgr/ha/etcd_tool.go]: s/activeKeyPreFIx/activeKeyPrefix

File: supernode/daemon/mgr/ha/etcd_tool.go [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+	activeKeyPreFIx = "/supernode/active/"
+	//standbyKeyPreFix is the keyPrefix of active supernode
```
Comment (Starnop) [inline supernode/daemon/mgr/ha/etcd_tool.go]: s/ of active supernode/ of standby supernode

File: supernode/daemon/mgr/ha/etcd_tool.go [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+	// build connection to etcd.
+	client, err := clientv3.New(config)
+	return &EtcdMgr{
+		hostIP: cfg.AdvertiseIP,
+		config: config,
+		client: client,
+	}, err
+}
+
+//WatchActiveChange is the progress to watch the etcd,if the value of key /lock/active changes,supernode will be notified.
```
Comment (Starnop) [inline supernode/daemon/mgr/ha/etcd_tool.go]: `/lock/active` or `/supernode/active/`? BTW, I think maybe `activeKeyPreFIx` is better.

File: supernode/daemon/mgr/ha/manager.go [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+import (
+	"context"
+
+	"github.com/dragonflyoss/Dragonfly/common/constants"
+	"github.com/dragonflyoss/Dragonfly/supernode/config"
+
+	"github.com/sirupsen/logrus"
+)
+
+//Manager is the struct to manager supernode ha.
```
Comment (Starnop) [inline supernode/daemon/mgr/ha/manager.go]: There should be a space between the contents and `//`.

File: supernode/daemon/mgr/ha_mgr.go [hunk 14]
```diff
@@ -0,0 +1,19 @@
+package mgr
+
+//HaMgr is the interface to implement supernode Ha.
+type HaMgr interface {
+	//ElectDaemonthe is the daemon progress to implement active/standby switch.
+	ElectDaemon(change chan int)
+
+	//HagetSupernodeState get supernode's status.
```
Comment (Starnop) [inline supernode/daemon/mgr/ha_mgr.go]: s/get/gets. The third person singular should be used here. So do with the other comments.

**Comment by Starnop** [comment]
LGTM.

@lowzj  PTAL.  THX.

File: supernode/daemon/mgr/ha/etcd_tool.go [hunk 15]
```diff
@@ -0,0 +1,164 @@
...
+		value = string(v.Value)
+	}
+	return value, err
+}
+
+// ActiveResureItsStatus keeps look on the lease's renew response.
+func (etcd *EtcdMgr) ActiveResureItsStatus() {
+	for {
+		select {
+		case keepResp := <-etcd.leaseKeepAliveRsp:
```
Comment (lowzj) [inline supernode/daemon/mgr/ha/etcd_tool.go]: It's very weird that `etcd.leaseKeepAliveRsp` is assigned in method `etcd.KeepAlive`. The channel `leaseKeepAliveRsp` maybe changed to `nil` in another goroutine when current goroutine is waiting for receiving from it. A receive from nil channel blocks forever.

File: supernode/daemon/mgr/ha/etcd_tool.go [hunk 16]
```diff
@@ -0,0 +1,164 @@
...
+
+	kv := clientv3.NewKV(etcd.client)
+
+	// make a lease to obtain a lock
+	lease := clientv3.NewLease(etcd.client)
+	if leaseResp, err = lease.Grant(ctx, etcd.leaseTTL); err != nil {
+		logrus.Errorf("failed to create a etcd lease: %v", err)
+	}
+
+	//send a heart beat to renew the lease
```
Comment (lowzj) [inline supernode/daemon/mgr/ha/etcd_tool.go]: leave a space between `//` and `send`

File: supernode/daemon/mgr/ha/etcd_tool.go [hunk 17]
```diff
@@ -0,0 +1,164 @@
...
+	}
+}
+
+// ObtainActiveInfo obtains the active supernode's information from etcd.
+func (etcd *EtcdMgr) ObtainActiveInfo(ctx context.Context, key string) (string, error) {
+	var value string
+	kv := clientv3.NewKV(etcd.client)
+	getRes, err := kv.Get(ctx, key)
+	if err != nil {
+		logrus.Errorf("failed to get the active supernode's(key: %s) info: %v", key, err)
```
Comment (lowzj) [inline supernode/daemon/mgr/ha/etcd_tool.go]: Why not return immediately when we get non-nil error?

**Comment by yunfeiyangbuaa** [comment]
@lowzj Thanks for your review.I have modify the code as your advice ,please review it again

File: dfget/core/api/supernode_api.go [hunk 18]
```diff
@@ -53,6 +53,8 @@ type SupernodeAPI interface {
 	ReportPiece(node string, req *types.ReportPieceRequest) (resp *types.BaseResponse, e error)
 	ServiceDown(node string, taskID string, cid string) (resp *types.BaseResponse, e error)
 	ReportClientError(node string, req *types.ClientErrorRequest) (resp *types.BaseResponse, e error)
+	Get(url string, resp interface{}) error
```
Comment (lowzj) [inline dfget/core/api/supernode_api.go]: Why does make the `get` and `post` method to public? They are not the apis of supernode.

**Comment by yunfeiyangbuaa** [comment]
all pr are  submitted ,please to review it

File: supernode/daemon/daemon.go [hunk 19]
```diff
@@ -81,6 +81,12 @@ func (d *Daemon) RegisterSuperNode() error {
 
 // Run runs the daemon.
 func (d *Daemon) Run() error {
+	if d.config.UseHA == true {
```
Comment (inoc603) [inline supernode/daemon/daemon.go]: `if d.config.UseHA {`

**Comment by jim3ma** [comment]
@yunfeiyanggzq how about implement supernode manager and supernode locator with etcd ?

**Comment by Starnop** [comment]
> @yunfeiyanggzq how about implement supernode manager and supernode locator with etcd ?

👍  also cc/ @lowzj

**Pull request closed without merging**